enhancements/authentication/configuring-webhook-token-authenticators.md

```diff
@@ -0,0 +1,169 @@
+  namespace, the name of the secret is derived from its index in the `spec.webhookTokenAuthenticators`
+  field of the `authentications.config/cluster` object such that `name=webhook-authenticator-<index>`
+3. if no webhook authenticators pass validation, the operator keeps the previous state
+4. redeploys the kube-apiserver static pods while specifying the new
+  mounted files in the `webhookTokenAuthenticators` [config field](https://github.com/openshift/api/blob/992bc3a41fe6c4e5bcf27dc8080f9e5c114c33fc/kubecontrolplane/v1/types.go#L146).
+
+#### Backend - Integrated OAuth Server Token Authentication
+
+1. cluster-authentication-operator observes `authentication.config/cluster` `type` field
+2. if set to "IntegratedOAuth", it creates an `openshift-config/webhook-authentication-integrated-oauth`
+  secret and adds it to the `webhookTokenAuthenticators` field of the `authentication.config/cluster`
+  resource
+  - the above secret contains details that point kube-apiserver to the oauth-apiserver for token reviews
+3. during an authentication flow, kube-apiserver sends `tokenreviews.authentication.k8s.io` object to the
+  oauth-apiserver for review
+4. oauth-apiserver sends a response accordingly to Kubernetes requirements, on success it adds the
+  `system:authenticated:oauth` group to the returned details about the groups of the authenticated user
+  - `system:authenticated:oauth` possesses privileges for users to handle projects via the self-provisioner
+    cluster role
+  - similarly, if any integrator would like their users to have project privileges, they need to add this
+    specific group in their token review responses.
+
+### Common Gotchas When Providing Your Own Authentication Webhook
+
+1. `oauth-proxy` relies on the _/.well-known/oauth-authorization-server_ endpoint
+   and if the integrated oauth-server is turned off and the `spec.oauthMetadata` field
+   in `authentications.config/cluster` is not set   properly, things like grafana
+   and prometheus won't be accessible
+2. ServiceAccounts may not be usable as OAuth clients anymore
+
+It is a sole responsibility of the replacing part to make sure all of these work.
+
+## Design Details
+
+### Test Plan
+
+Since we will be making the cluster-authentication-operator use the `webhookTokenAuthenticatros`
+(in combination with `oauthMetadata`, which already works) to configure authentication using
+the integrated OAuth server, we will just re-use all the tests that we have already.
+
+### Upgrade / Downgrade Strategy
+
+To keep authentication working even during upgrades, we'll keep our kube-apiserver
+authenticator patch code in place for 4.n version, but will disable it if any webhook
+authenticators are set. The authenticator patching code should then be removed in 4.(n+1)
+version.
+
+The cluster-authentication-operator in 4.n will wait for the kube-apiserver-operator to
+announce its 4.n version, and once that happens, and only after the oauth-apiserver deployment
```

Comment:
how does it announce that?

Reply:
Using the `versions` field of the `clusteroperator/kube-apiserver`

Reply:
> Using the `versions` field of the `clusteroperator/kube-apiserver`

We can't use this.  The versions aren't supposed to be relied upon.  @p0lyn0mial had a way to do this with conditions.

Reply:
correct, follow https://github.com/openshift/api/blob/master/operator/v1/types_authentication.go#L31

Reply:
I'll look at how this is handled, thanks for the pointers

Reply:
actually, now that I think of it, if 4.(n-1) version gets the oauth-apiserver, in 4.n we may just wait for the oauth-apiserver deployment to become available and simply switch to webhook authenticators from the authn operator.

4.(n-1) ignores the webhook authenticator config (I'll add the flag configuration to unsupportedConfigOverrides so that we can offer something to the Keycloak people and me testing the oauth-apiserver tokenvalidation endpoint), it only gets properly observed in 4.n.

Both 4.(n-1) and 4.n will keep the authentication code in KAS, it will only get used if no webhook authenticators are configured, it will get removed in 4.(n+1). That should keep all paths safe of authentication disruptions I think.

Let me add that to the doc.

Reply:
And I see that's more or less what was there before with the exception to having to wait for the KAS, which should perhaps not be necessary if the oauth-apiserver is already in place?

Anyway, added a small bit, hopefully it makes sense.